numba/np/npyimpl.py

```diff
@@ -411,6 +492,64 @@ def numpy_ufunc_kernel(context, builder, sig, args, ufunc, kernel_class):
     return impl_ret_new_ref(context, builder, sig.return_type, out)
 
 
+def numpy_gufunc_kernel(context, builder, sig, args, ufunc, kernel_class):
+    arguments = []
+    expected_ndims = kernel_class.dufunc.expected_ndims()
+    expected_ndims = expected_ndims[0] + expected_ndims[1]
+    is_input = [True] * ufunc.nin + [False] * ufunc.nout
+    for arg, ty, exp_ndim, is_inp in zip(args, sig.args, expected_ndims, is_input):  # noqa: E501
+        if isinstance(ty, types.ArrayCompatible):
+            # Create an array helper that iteration returns a subarray
+            # with ndim specified by "exp_ndim"
+            arr = context.make_array(ty)(context, builder, arg)
+            shape = cgutils.unpack_tuple(builder, arr.shape, ty.ndim)
+            strides = cgutils.unpack_tuple(builder, arr.strides, ty.ndim)
+            inner_arr_ty = ty.copy(ndim=exp_ndim)
+            ndim = ty.ndim
+            layout = ty.layout
+            base_type = ty.dtype
+            array_helper = _ArrayGUHelper(context, builder,
+                                          shape, strides, arg,
+                                          layout, base_type, ndim,
+                                          inner_arr_ty, is_inp)
```

Comment:
This needs to check that the input arrays' `.shape` are matching the `loopshape` (which is defined later by the output array). 

Currently, a mismatched input shape will result in incorrect behavior:

```python
from numba import jit, guvectorize
import numpy as np


@guvectorize('() -> ()')
def gufunc(x, res):
    res[0] = x + 100

@jit(nopython=True)
def jit_func(x, res):
    gufunc(x, res)

x = np.arange(40, dtype='i8').reshape(1, 5, 2, 4)
res = np.zeros((5, 2, 4), dtype=x.dtype)
expected = np.zeros_like(res)

print('----jit-func')
jit_func(x, res)
print(res)
# Out:
# [[[100 100 100 100]
#   [100 100 100 100]]

#  [[100 100 100 100]
#   [100 100 100 100]]

#  [[100 100 100 100]
#   [100 100 100 100]]

#  [[100 100 100 100]
#   [100 100 100 100]]

#  [[100 100 100 100]
#   [100 100 100 100]]]

print('----gfunc')


gufunc(x,  expected)
print(expected)
# Out:
# [[[100 101 102 103]
#   [104 105 106 107]]

#  [[108 109 110 111]
#   [112 113 114 115]]

#  [[116 117 118 119]
#   [120 121 122 123]]

#  [[124 125 126 127]
#   [128 129 130 131]]

#  [[132 133 134 135]
#   [136 137 138 139]]]
```

I just realize the same problem exist for `numpy_ufunc_kernel` above:

```python
from numba import jit, vectorize
import numpy as np


@vectorize
def ufunc(x):
    return x + 100

@jit(nopython=True)
def jit_func(x, res):
    ufunc(x, res)

x = np.arange(40, dtype='i8').reshape(1, 5, 2, 4)
res = np.zeros((5, 2, 4), dtype=x.dtype)
expected = np.zeros_like(res)

print('----jit-func')
jit_func(x, res)
print(res)
# Out:
# [[[100 100 100 100]
#   [100 100 100 100]]

#  [[100 100 100 100]
#   [100 100 100 100]]

#  [[100 100 100 100]
#   [100 100 100 100]]

#  [[100 100 100 100]
#   [100 100 100 100]]

#  [[100 100 100 100]
#   [100 100 100 100]]]

print('----ufunc')


ufunc(x,  expected)
print(expected)
# Out:
# [[[100 101 102 103]
#   [104 105 106 107]]

#  [[108 109 110 111]
#   [112 113 114 115]]

#  [[116 117 118 119]
#   [120 121 122 123]]

#  [[124 125 126 127]
#   [128 129 130 131]]

#  [[132 133 134 135]
#   [136 137 138 139]]]
```

Reply:
@guilhermeleobas 

Here's a rough starter patch to do the shape checking:

```diff
diff --git a/numba/np/npyimpl.py b/numba/np/npyimpl.py
index 1c1bc6e2a..8581b1975 100644
--- a/numba/np/npyimpl.py
+++ b/numba/np/npyimpl.py
@@ -7,6 +7,7 @@ import math
 import sys
 import itertools
 from collections import namedtuple
+from functools import reduce
 
 import llvmlite.ir as ir
 
@@ -236,6 +237,14 @@ class _ArrayGUHelper(namedtuple('_ArrayHelper', ('context', 'builder',
                                          index_types, indices)
             return impl_ret_borrowed(context, builder, self.inner_arr_ty, res)
 
+    def guard_shape(self, loopshape):
+        context, builder = self.context, self.builder
+        matches = [builder.icmp_unsigned('==', sh, ex) for sh, ex in zip(self.shape, loopshape)]
+        all_match = reduce(builder.and_, matches)
+
+        with builder.if_then(builder.not_(all_match)):
+            context.call_conv.return_user_exc(builder, ValueError, ("shape incompatible",))
+
 
 def _prepare_argument(ctxt, bld, inp, tyinp, where='input operand'):
     """returns an instance of the appropriate Helper (either
@@ -538,6 +547,10 @@ def numpy_gufunc_kernel(context, builder, sig, args, ufunc, kernel_class):
     loopshape_ndim = outputs[0].ndim - outputs[0].inner_arr_ty.ndim
     loopshape = outputs[0].shape[ : loopshape_ndim]
 
+
+    for arg in arguments[:ufunc.nin]:
+        if isinstance(arg, _ArrayGUHelper):
+            arg.guard_shape(loopshape)
     with cgutils.loop_nest(builder,
                            loopshape,
                            intp=intpty,

```

Reply:
Thanks, Siu! Commit c029cd5 contains two checks:
* Shape mismatch
* Arguments with same signature should have the same core dimensions

I also wrote the code to raise the exception using `context.compile_internal` as it is easier to maintain and review Python code than LLVM IR.